calico_kubernetes/calico_kubernetes.py

```diff
@@ -141,8 +128,233 @@ def _configure_interface(self):
+        :param ip: An IPAddress object with the desired IP to assign.
+        :param interface: The name of the interface in the container.
+        """
+        # Check if the container already exists. If it does, exit.
+        try:
+            _ = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=orchestrator_id,
+                workload_id=workload_id
+            )
+        except KeyError:
+            # Calico doesn't know about this container.  Continue.
+            pass
+        else:
+            # Calico already set up networking for this container.  Since we got
+            # called with an IP address, we shouldn't just silently exit, since
+            # that would confuse the user: the container would not be reachable on
+            # that IP address.
+            print("This container has already been configured with Calico Networking.")
+            sys.exit(1)
+
+        # Obtain information from Docker Client and validate container state
+        self._validate_container_add(container_name)
+
+        # Obtain ip pool from datastore client, validate pool, and assign
+        # ip address through IPAM Client
+        self._assign_addresses([ip])
+
+        # Obtain next hops from datastore client, validate next hop, and
+        # create Endpoint object
+        ep = self._create_endpoint(workload_id, ip, hostname, orchestrator_id)
+
+        # Create the veth, move into the container namespace, add the IP and
+        # set up the default routes.
+        # Update Endpoint in felix
+        ep = self._create_veth(ep, pid, ip, interface)
+
+        # Let the caller know what endpoint was created.
+        return ep
+
+    def _container_remove(self, container_name, workload_id, hostname, orchestrator_id):
+        """
+        Remove the indicated container on this host from Calico networking
+
+        :param container_name: THe name or partial id of the container
+        """
+        # Find the endpoint ID. We need this to find any ACL rules
+        try:
+            endpoint = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=orchestrator_id,
+                workload_id=workload_id
+            )
+        except KeyError:
+            print("Container %s doesn't contain any endpoints" % container_name)
+            sys.exit(1)
+
+        # Remove any IP address assignments that this endpoint has
+        for net in endpoint.ipv4_nets | endpoint.ipv6_nets:
+            assert(net.size == 1)
+            ip = net.ip
+            self._unassign_addresses([ip])
+
+        # Remove the endpoint
+        netns.remove_veth(endpoint.name)
+
+        # Remove the container from the datastore.
+        self._datastore_client.remove_workload(hostname, orchestrator_id, workload_id)
+
+        print("Removed Calico interface from %s" % container_name)
+
+    def _validate_container_add(self, container_name):
```

Comment:
A better name for this method is probably something like `_validate_container_state` - it isn't specifically tied to the container_add logic at all.